stan/math/prim/mat/prob/bernoulli_logit_glm_lpmf.hpp

```diff
@@ -81,63 +89,63 @@ typename return_type<T_x, T_alpha, T_beta>::type bernoulli_logit_glm_lpmf(
   if (!include_summand<propto, T_x, T_alpha, T_beta>::value)
     return 0.0;
 
-  Matrix<T_partials_return, Dynamic, 1> signs(N, 1);
-  {
-    scalar_seq_view<T_y> y_vec(y);
-    for (size_t n = 0; n < N; ++n) {
-      signs[n] = 2 * y_vec[n] - 1;
-    }
-  }
-  Matrix<T_partials_return, Dynamic, 1> beta_dbl(M, 1);
-  {
-    scalar_seq_view<T_beta> beta_vec(beta);
-    for (size_t m = 0; m < M; ++m) {
-      beta_dbl[m] = value_of(beta_vec[m]);
-    }
-  }
-  Eigen::Array<T_partials_return, Dynamic, 1> ytheta
-      = signs.array() * (value_of(x) * beta_dbl).array();
-  scalar_seq_view<T_alpha> alpha_vec(alpha);
+  const auto &x_val = value_of_rec(x);
+  const auto &y_val = value_of_rec(y);
+  const auto &beta_val = value_of_rec(beta);
+  const auto &alpha_val = value_of_rec(alpha);
+
+  const auto &y_val_vec = as_column_vector_or_scalar(y_val);
+  const auto &beta_val_vec = as_column_vector_or_scalar(beta_val);
+  const auto &alpha_val_vec = as_column_vector_or_scalar(alpha_val);
+
+  T_y_val signs = 2 * as_array_or_scalar(y_val_vec) - 1;
+
+  Eigen::Array<T_partials_return, Dynamic, 1> ytheta = x_val * beta_val_vec;
+  ytheta = as_array_or_scalar(signs)
+           * (ytheta + as_array_or_scalar(alpha_val_vec));
 
   // Compute the log-density and handle extreme values gracefully
   // using Taylor approximations.
   // And compute the derivatives wrt theta.
   static const double cutoff = 20.0;
-  Matrix<T_partials_return, Dynamic, 1> theta_derivative(N, 1);
-  T_partials_return theta_derivative_sum = 0;
-  T_partials_return exp_m_ythetan;
-  for (size_t n = 0; n < N; ++n) {
-    ytheta[n] += signs[n] * value_of(alpha_vec[n]);
-    exp_m_ythetan = exp(-ytheta[n]);
-    if (ytheta[n] > cutoff) {
-      logp -= exp_m_ythetan;
-      theta_derivative[n] = -exp_m_ythetan;
-    } else if (ytheta[n] < -cutoff) {
-      logp += ytheta[n];
-      theta_derivative[n] = signs[n];
-    } else {
-      logp -= log1p(exp_m_ythetan);
-      theta_derivative[n] = signs[n] * exp_m_ythetan / (exp_m_ythetan + 1);
+  Eigen::Array<T_partials_return, Dynamic, 1> exp_m_ytheta = exp(-ytheta);
+  logp += (ytheta > cutoff)
+              .select(-exp_m_ytheta,
+                      (ytheta < -cutoff).select(ytheta, -log1p(exp_m_ytheta)))
+              .sum();
+  if (!std::isfinite(logp)) {
+    check_finite(function, "Weight vector", beta);
+    check_finite(function, "Intercept", alpha);
+    for (size_t n = 0; n < N; ++n) {
```

Comment:
There should be a check-finite that applies to containers that will return more meaningful messages.

Reply:
Can you suggest what the error message should be? I just left them as they were.

Reply:
Sorry that wasn't clearer.  The message doesn't need to change, it just shouldn't be in a loop.  We shold have `check_` methods that work on containers so that they can indicate the index(es) at which the failure occurs.  The problem with having a loop here like this is that we don't indicate which element failed.

This also isn't a big deal if the container tests don't (yet) exist, but we should be creating issues for them so we can get rid of these loops.